ServerCore/Helpers/SubmissionEvaluator.cs

```diff
@@ -0,0 +1,356 @@
+            {
+                submission.UnformattedSubmissionText = submissionText;
+            }
+            else
+            {
+                submission.SubmissionText = submissionText;
+            }
+
+            submission.Response = await context.Responses.Where(
+                r => r.Puzzle.ID == puzzleId &&
+                     submissionTextToCheck == r.SubmittedText)
+                .FirstOrDefaultAsync();
+
+            submissions.Add(submission);
+
+            // Update puzzle state if submission was correct
+            if (submission.Response != null && submission.Response.IsSolution)
+            {
+                await PuzzleStateHelper.SetSolveStateAsync(context,
+                    thisEvent,
+                    submission.Puzzle,
+                    submission.Team,
+                    submission.TimeSubmitted);
+            }
+            else if (!puzzle.IsFreeform && submission.Response == null && thisEvent.IsAnswerSubmissionActive)
+            {
+                // We also determine if the puzzle should be set to email-only mode.
+                if (IsPuzzleSubmissionLimitReached(
+                        thisEvent,
+                        submissions,
+                        puzzleState))
+                {
+                    await PuzzleStateHelper.SetEmailOnlyModeAsync(context,
+                        thisEvent,
+                        submission.Puzzle,
+                        submission.Team,
+                        true);
+
+                    var authors = await context.PuzzleAuthors.Where((pa) => pa.Puzzle == submission.Puzzle).Select((pa) => pa.Author.Email).ToListAsync();
+
+                    MailHelper.Singleton.SendPlaintextBcc(authors,
+                        $"{thisEvent.Name}: Team {submission.Team.Name} is in email mode for {submission.Puzzle.Name}",
+                        "");
+                }
+                else
+                {
+                    // If the submission was incorrect and not a partial solution,
+                    // we will do the lockout computations now.
+                    DateTime? lockoutExpiryTime = ComputeLockoutExpiryTime(
+                        thisEvent,
+                        submissions,
+                        puzzleState);
+
+                    if (lockoutExpiryTime != null)
+                    {
+                        await PuzzleStateHelper.SetLockoutExpiryTimeAsync(context,
+                            thisEvent,
+                            submission.Puzzle,
+                            submission.Team,
+                            lockoutExpiryTime);
+                    }
+                }
+            }
+
+            context.Submissions.Add(submission);
+            await context.SaveChangesAsync();
+
+            if (submission.Response != null && submission.Response.IsSolution)
+            {
+                return new SubmissionResponse() { ResponseCode = SubmissionResponseCode.Correct };

```

Comment:
Shouldn't correct submission still have the response set (e.g. "Correct! The meta clues are in Building 7")